tools/ml-dataprep/ml_dataprep/dataprep.py

```diff
@@ -0,0 +1,244 @@
+                                                self._destination_dataset,
+                                                table_id,
+                                                destination_uri)
+        except GoogleCloudError as gcp_exception:
+            raise exceptions.AppException('Could not generate {} dataset'.format(ml_dataset),
+                                          ERR_GENERATE_ML_DATASET,
+                                          gcp_exception)
+
+    def _extract_training_dataset(self, temp_table_uri, timestamp, split_index):
+        """"Extract the training dataset from the shuffled temporary table. """
+        self._extract_ml_dataset(TRAINING_DATASET,
+                                 temp_table_uri,
+                                 timestamp,
+                                 split_index)
+
+    def _extract_validation_dataset(self, temp_table_uri, timestamp, split_index):
+        """"Extract the validation dataset from the shuffled temporary table."""
+        self._extract_ml_dataset(VALIDATION_DATASET,
+                                 temp_table_uri,
+                                 timestamp,
+                                 split_index)
+
+    def extract_all_ml_datasets(self):
+        """Extract the ML training and validation datasets from BigQuery.
+
+        Extract the ML training and validation datasets from BigQuery
+        in the following sequence of steps:
+            1) Shuffle the source data and save it in a temporary
+            table.
+            2) Calculate the split index where to separate training and
+            validation data.
+            3) Extract training and validation data from the temporary
+            table into separate BigQuery tables.
+            4) Export the training and validation data BigQuery tables
+            in CSV format to Cloud Storage.
+        """
+        temp_table_uri = None
+        try:
+            total_lines, split_index = self._calculate_dataset_sizes()
+            timestamp = datetime.datetime.today().strftime(TIMESTAMP_FORMAT)
+            temp_table_uri = self._create_temp_table(timestamp, total_lines)
+            self._extract_training_dataset(
+                temp_table_uri, timestamp, split_index)
+            self._extract_validation_dataset(
+                temp_table_uri, timestamp, split_index)
+        except exceptions.AppException as app_exception:
+            logging.error(app_exception)
+            if app_exception.code <= ERR_CREATE_TEMP_TABLE:
+                # Don't need to execute finally block, temp table not created
+                return
+        finally:
+            if temp_table_uri is not None:
+                logging.info('Deleting temporary table %s', temp_table_uri)
+                try:
+                    self._bq_client.delete_table(temp_table_uri)
+                except Exception:
+                    # Ignore exceptions since we're in the finally block
+                    pass
```

Comment:
1. don't `pass` on the final exception, add a warning at least
2. do you really need nested `try/except` logic? It's probably good to have exceptions raised in `finally` block because users may need to delete temp tables manually.

Reply:
You are right, I was thinking the same thing 5 minutes ago:) I removed the try/except in the finally in 42df0bf906677b04a849cd4edfa755cbd8062d8a